kubectl/upload/main.go

```diff
@@ -0,0 +1,212 @@
+			dynamicClient, err := dynamic.NewForConfig(config)
+			if err != nil {
+				log.Fatal(err)
+			}
+
+			client := &realKubernetesClient{
+				dynamicClient: dynamicClient,
+			}
+
+			err = run(cfg, client)
+			if err != nil {
+				log.Fatal(err)
+			}
+		},
+	}
+	cmd.Flags().StringVarP(&cfg.Path, "path", "p", ".", "Path to the directory to be uploaded")
+	cmd.Flags().StringVarP(&cfg.Resource.Namespace, "namespace", "n", "default", "Namespace of the resource created")
+	if home := homeDir(); home != "" {
+		cmd.Flags().StringVarP(&cfg.Kubeconfig, "kubeconfig", "", filepath.Join(home, ".kube", "config"), "")
+	} else {
+		cmd.Flags().StringVarP(&cfg.Kubeconfig, "kubeconfig", "", "", "")
+	}
+	cmd.Flags().BoolVarP(&cfg.Verbose, "verbose", "v", false, "Verbose output")
+
+	if err := cmd.Execute(); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func run(cfg Config, client KubernetesClient) error {
+	if !fileExists(filepath.Join(cfg.Path, "Dockerfile")) {
+		return errors.New("a Dockerfile does not exist at the given path")
+	}
+
+	if cfg.Verbose {
+		fmt.Println("packaging the directory into a tarball")
+	}
+
+	tarPath := "/tmp/archive.tar.gz"
+	err := tarGz(cfg.Path, tarPath)
+	if err != nil {
+		return err
+	}
+	defer os.Remove(tarPath)
+
+	checksum, err := calculateMD5(tarPath)
+	if err != nil {
+		return err
+	}
+
+	cfg.Resource.Md5Checksum = checksum
+	data, err := hex.DecodeString(checksum)
+	if err != nil {
+		return err
+	}
+	cfg.Resource.EncodedMd5 = base64.StdEncoding.EncodeToString(data)
+
+	customResource := createCustomResource(cfg)
+
+	gvr := schema.GroupVersionResource{
+		Group:    "substratus.ai",
+		Version:  "v1",
+		Resource: strings.ToLower(cfg.Resource.Kind) + "s", // plural is needed here
+	}
+
+	result, err := client.CreateResource(gvr, cfg.Resource.Namespace, customResource, metav1.CreateOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to create custom resource: %w", err)
+	}
+
+	watcher, err := client.WatchResource(gvr, cfg.Resource.Namespace, metav1.ListOptions{
+		FieldSelector: fmt.Sprintf("metadata.name=%s", customResource.GetName()),
+	})
+	if err != nil {
+		return fmt.Errorf("failed to watch resource %q: %w", customResource.GetName(), err)
+	}
+	if cfg.Verbose {
+		fmt.Printf("Created custom resource %q.\n", result.GetName())
+	}
+
+	// Monitor the watcher channel
+	for event := range watcher.ResultChan() {
+		switch event.Type {
+		case watch.Added, watch.Modified:
+			err = handleWatchEvent(event, cfg, tarPath)
+			if err != nil {
+				return err
+			}
+		case watch.Error:
+			return errors.New("encountered a watch error")
+		case watch.Deleted:
+			return errors.New("the custom resource was deleted")
+		default:
+			return errors.New("unhandled event type")
+		}
+	}
+
+	if cfg.Verbose {
+		fmt.Printf("Upload is complete. Waiting for the build job to complete.\n")
+	}
+	// TODO(bjb): should we put a watcher also on name-kind-container-builder or does the utility stop here?
```

Comment:
I'm going to follow through on this last bit because it applies to any resource that we're building. I think if a build fails, it's worth surfacing that to the client so it's not a complete mystery what happened on the in-cluster build.

Reply:
This info should be surfaced from the `.status.condtions` set by the container image reconciler